internal/application/manager.go

```diff
@@ -51,11 +65,92 @@ func NewManager(service appInterfaces.ApplicationService) interfaces.DataManager
 // StartRecording starts a recording session based on the values in the request.
 // An error is returned if the request data is incomplete or a record or replay session is currently running.
 func (m *dataManager) StartRecording(request *dtos.RecordRequest) error {
-	//TODO implement me using TDD
+	lc := m.appSvc.LoggingClient()
 
-	// TODO: Use new appSrv.AppContext.Done() to exit from long-running recording function
+	m.recordingMutex.Lock()
+	defer m.recordingMutex.Unlock()
 
-	return errors.New("not implemented")
+	if m.recordingStartedAt != nil {
+		return recordingInProgressError
+	}
+	m.recordedData = nil
+	m.eventCount = 0
+
+	var pipeline []appInterfaces.AppFunction
+
+	if len(request.IncludeDeviceProfiles) > 0 {
+		includeFilter := transforms.NewFilterFor(request.IncludeDeviceProfiles)
+		pipeline = append(pipeline, includeFilter.FilterByProfileName)
+		lc.Debugf(debugFilterMessage, "for profile", request.IncludeDeviceProfiles)
+	}
+
+	if len(request.ExcludeDeviceProfiles) > 0 {
+		excludeFilter := transforms.NewFilterOut(request.ExcludeDeviceProfiles)
+		pipeline = append(pipeline, excludeFilter.FilterByProfileName)
+		lc.Debugf(debugFilterMessage, "out profile", request.ExcludeDeviceProfiles)
+	}
+
+	if len(request.IncludeDevices) > 0 {
+		includeFilter := transforms.NewFilterFor(request.IncludeDevices)
+		pipeline = append(pipeline, includeFilter.FilterByDeviceName)
+		lc.Debugf(debugFilterMessage, "for device", request.IncludeDevices)
+	}
+
+	if len(request.ExcludeDevices) > 0 {
+		excludeFilter := transforms.NewFilterOut(request.ExcludeDevices)
+		pipeline = append(pipeline, excludeFilter.FilterByDeviceName)
+		lc.Debugf(debugFilterMessage, "out device", request.ExcludeDevices)
+	}
+
+	if len(request.IncludeSources) > 0 {
+		includeFilter := transforms.NewFilterFor(request.IncludeSources)
+		pipeline = append(pipeline, includeFilter.FilterBySourceName)
+		lc.Debugf(debugFilterMessage, "for source", request.IncludeSources)
+	}
+
+	if len(request.ExcludeSources) > 0 {
+		excludeFilter := transforms.NewFilterOut(request.ExcludeSources)
+		pipeline = append(pipeline, excludeFilter.FilterBySourceName)
+		lc.Debugf(debugFilterMessage, "out source", request.ExcludeSources)
+	}
+
+	var batch *transforms.BatchConfig
+	var err error
+
+	if request.Duration > 0 && request.EventLimit > 0 {
+		batch, err = transforms.NewBatchByTimeAndCount(request.Duration.String(), request.EventLimit)
+	} else if request.EventLimit > 0 {
+		batch, err = transforms.NewBatchByCount(request.EventLimit)
+	} else if request.Duration > 0 {
+		batch, err = transforms.NewBatchByTime(request.Duration.String())
+	}
+
+	if batch == nil {
+		err = batchParametersNotSetError
+	}
+
+	if err != nil {
+		return fmt.Errorf("%s: %v", createBatchFailedMessage, err)
+	}
```

Comment:
Shouldn't the checking order of batch and err reverse?   i.e. if there is an error then you use that error. otherwise if no error then check the batch?

Reply:
Good catch , that error should be in `else` clause w/o conditional and that will clear it up.